services/mirror/mirror.go

```diff
@@ -78,21 +82,56 @@ func Update(ctx context.Context) error {
 			return nil
 		}
 
+		// Check we've not been cancelled
 		select {
 		case <-ctx.Done():
-			return fmt.Errorf("Aborted")
+			return fmt.Errorf("aborted")
 		default:
-			return mirrorQueue.Push(&item)
 		}
+
+		// Check if this request is already in the queue
+		has, err := mirrorQueue.Has(&item)
+		if err != nil {
+			return err
+		}
+		if has {
+			return nil
+		}
+
+		// Double check we've not been cancelled
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("aborted")
+		default:
+		}
```

Comment:
This check doesn't seem to help much. Because the context can still be canceled after this check and then `mirrorQueue.Push` gets called.

So maybe we can just keep the first `check canceled` only.

Reply:
It's purpose is to check if the context has been cancelled during the Has. 

What is the problem of checking again? Why would you want to not check again?

Reply:
The basic question in my mind is "what is the purpose of checking again, what's the benefit of checking again."

The idea comes from Occam's razor  https://en.wikipedia.org/wiki/Occam's_razor 

If the results are equivalent, maybe it's better to keep the simpler one.

About the checks:

1. With the second check, there are still many  chances that the context would be canceled after the second check, the code works just like the second check doesn't exist.
2. Without the second check, the first check has already handled the context cancelling correctly.
3. If there is the reason we need the second check, then do we need a third and forth checks? eg: check before `requested++`, check before `return`, check between `MirrorsIterate-pull` and `MirrorsIterate-push`, etc.